domain/secret/service/access.go

```diff
@@ -237,15 +228,16 @@ func (s *SecretService) getManagementCaveat(
 				// Typically not found error.
 				return nil, errors.Trace(err)
 			}
+
 			if hasRole.Allowed(secrets.RoleManage) {
 				return func(ctx context.Context, fn func(context.Context) error) error {
-					return fn(ctx)
+					return s.leaderEnsurer.WithLeader(ctx, appName, accessor.ID, fn)
```

Comment:
This is the material change. Where an application-owned secret requires the unit accessor to be the leader, the returned caveat guarantees this condition for the duration of the call.